File: .github/actions-scripts/fr-add-docs-reviewers-requests.js
import { graphql } from '@octokit/graphql'

import {
  addItemsToProject,
  isDocsTeamMember,
  findFieldID,
  findSingleSelectID,
  generateUpdateProjectNextItemFieldMutation,
} from './projects.js'

async function run() {
  // Get info about the docs-content review board project
  // and about open github/github PRs
  const data = await graphql(
    `
      query ($organization: String!, $repo: String!, $projectNumber: Int!, $num_prs: Int!) {
        organization(login: $organization) {
          projectNext(number: $projectNumber) {
            id
            items(last: 100) {
              nodes {
                id
              }
            }
            fields(first: 20) {
              nodes {
                id
                name
                settings
              }
            }
          }
        }
        repository(name: $repo, owner: $organization) {
          pullRequests(last: $num_prs, states: OPEN) {
            nodes {
              id
              isDraft
              reviewRequests(first: 10) {
                nodes {
                  requestedReviewer {
                    ... on Team {
                      name
                    }
                  }
                }
              }
              labels(first: 5) {
                nodes {
                  name
                }
              }
              reviews(first: 10) {
                nodes {
                  onBehalfOf(first: 1) {
                    nodes {
                      name
                    }
                  }
                }
              }
              author {
                login
              }
            }
          }
        }
      }
    `,
    {
      organization: process.env.ORGANIZATION,
      repo: process.env.REPO,
      projectNumber: parseInt(process.env.PROJECT_NUMBER),
      num_prs: parseInt(process.env.NUM_PRS),
      headers: {
        authorization: `token ${process.env.TOKEN}`,
        'GraphQL-Features': 'projects_next_graphql',
      },
    }
  )

  // Get the PRs that are:
  // - not draft
  // - not a train
  // - are requesting a review by docs-reviewers
  // - have not already been reviewed on behalf of docs-reviewers
  const prs = data.repository.pullRequests.nodes.filter(
    (pr) =>
      !pr.isDraft &&
      !pr.labels.nodes.find((label) => label.name === 'Deploy train 🚂') &&
      pr.reviewRequests.nodes.find(
        (requestedReviewers) => requestedReviewers.requestedReviewer.name === process.env.REVIEWER
      ) &&
      !pr.reviews.nodes
        .flatMap((review) => review.onBehalfOf.nodes)
        .find((behalf) => behalf.name === process.env.REVIEWER)
  )
  if (prs.length === 0) {
    console.log('No PRs found. Exiting.')
    return
  }

  const prIDs = prs.map((pr) => pr.id)
  const prAuthors = prs.map((pr) => pr.author.login)
  console.log(`PRs found: ${prIDs}`)

  // Get the project ID
  const projectID = data.organization.projectNext.id

  // Get the IDs of the last 100 items on the board.
  // Until we have a way to check from a PR whether the PR is in a project,
  // this is how we (roughly) avoid overwriting PRs that are already on the board.
  // If we are overwriting items, query for more items.
  const existingItemIDs = data.organization.projectNext.items.nodes.map((node) => node.id)

  // Get the ID of the fields that we want to populate
  const datePostedID = findFieldID('Date posted', data)
  const reviewDueDateID = findFieldID('Review due date', data)
  const statusID = findFieldID('Status', data)
  const featureID = findFieldID('Feature', data)
  const contributorTypeID = findFieldID('Contributor type', data)
  const authorID = findFieldID('Author', data)

  // Get the ID of the single select values that we want to set
  const readyForReviewID = findSingleSelectID('Ready for review', 'Status', data)
  const hubberTypeID = findSingleSelectID('Hubber or partner', 'Contributor type', data)
  const docsMemberTypeID = findSingleSelectID('Docs team', 'Contributor type', data)

  // Add the PRs to the project
  const itemIDs = await addItemsToProject(prIDs, projectID)

  // If an item already existed on the project, the existing ID will be returned.
  // Exclude existing items going forward.
  // Until we have a way to check from a PR whether the PR is in a project,
  // this is how we (roughly) avoid overwriting PRs that are already on the board
  const newItemIDs = itemIDs.filter((id) => !existingItemIDs.includes(id))

  if (newItemIDs.length === 0) {
    console.log('All found PRs are already on the project. Exiting.')
    return
  }

  // Populate fields for the new project items
  // (Using for...of instead of forEach since the function uses await)
  for (const [index, itemID] of newItemIDs.entries()) {
    const updateProjectNextItemMutation = generateUpdateProjectNextItemFieldMutation({
      item: itemID,
      author: prAuthors[index],
      turnaround: 2,
    })
    const contributorType = isDocsTeamMember(prAuthors[index]) ? docsMemberTypeID : hubberTypeID
    console.log(`Populating fields for item: ${itemID}`)

    await graphql(updateProjectNextItemMutation, {
      project: projectID,
      statusID: statusID,
      statusValueID: readyForReviewID,
      datePostedID: datePostedID,
      reviewDueDateID: reviewDueDateID,
      contributorTypeID: contributorTypeID,
      contributorType: contributorType,
      featureID: featureID,
      authorID: authorID,
      headers: {
        authorization: `token ${process.env.TOKEN}`,
        'GraphQL-Features': 'projects_next_graphql',
      },
    })
    console.log('Done populating fields for item')
  }

  return newItemIDs
}

run().catch((error) => {
  console.log(`#ERROR# ${error}`)
  process.exit(1)
})

File: .github/actions-scripts/fr-add-docs-reviewers-requests.py
# TODO: Convert to JavaScript for language consistency

import json
import logging
import os
import requests

# Constants
endpoint = 'https://api.github.com/graphql'

# ID of the github/github repo
github_repo_id = "MDEwOlJlcG9zaXRvcnkz"

# ID of the docs-reviewers team
docs_reviewers_id = "MDQ6VGVhbTQzMDMxMzk="

# ID of the "Docs content first responder" board
docs_project_id = "MDc6UHJvamVjdDQ1NzI0ODI="

# ID of the "OpenAPI review requests" column on the  "Docs content first responder" board
docs_column_id = "PC_lAPNJr_OAEXFQs4A2OFq"

# 100 is an educated guess of how many PRs are opened in a day on the github/github repo
# If we are missing PRs, either increase this number or increase the frequency at which this script is run
num_prs_to_search = 100

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def find_open_prs_for_repo(repo_id: str, num_prs: int):
  """Return data about a specified number of open PRs for a specified repo

  Arguments:
  repo_id: The node ID of the repo to search
  num_prs: The max number of PRs to return

  Returns:
  Returns a JSON object of this structure:
  {
    "data": {
    "node": {
      "pullRequests": {
        "nodes": [
            {
              "id": str, 
              "isDraft": bool, 
              "reviewRequests": {
                "nodes": [
                  {
                    "requestedReviewer": {
                      "id": str
                    }
                  }...
                ]
              }, 
              "projectCards": {
                "nodes": [
                  {
                    "project": {
                      "id": str
                    }
                  }...
                ]
              }
            }...
          ]
        }
      }
    }
  }
  """

  query = """query ($repo_id: ID!, $num_prs: Int!) {
    node(id: $repo_id) {
      ... on Repository {
        pullRequests(last: $num_prs, states: OPEN) {
          nodes {
            id
            isDraft
            reviewRequests(first: 10) {
              nodes {
                requestedReviewer {
                  ... on Team {
                    id
                  }
                }
              }
            }
            projectCards(first: 10) {
              nodes {
                project {
                  id
                }
              }
            }
          }
        }
      }
    }
  }
  """

  variables = {
    "repo_id": github_repo_id,
    "num_prs": num_prs
  }

  response = requests.post(
    endpoint, 
    json={'query': query, 'variables': variables}, 
    headers = {'Authorization': f"bearer {os.environ['TOKEN']}"}
    )

  response.raise_for_status()

  json_response = json.loads(response.text)

  if 'errors' in json_response:
    raise RuntimeError(f'Error in GraphQL response: {json_response}')

  return json_response

def add_prs_to_board(prs_to_add: list, column_id: str):
  """Adds PRs to a column of a project board

  Arguments:
  prs_to_add: A list of PR node IDs
  column_id: The node ID of the column to add the PRs to

  Returns:
  Nothing
  """

  logger.info(f"adding: {prs_to_add}")

  mutation = """mutation($pr_id: ID!, $column_id: ID!) {
                  addProjectCard(input:{contentId: $pr_id, projectColumnId: $column_id}) {
                    projectColumn {
                      name
                      }
                    }
  }"""

  for pr_id in prs_to_add:
    logger.info(f"Attempting to add {pr_id} to board")

    variables = {
      "pr_id": pr_id,
      "column_id": column_id
    }

    response = requests.post(
      endpoint, 
      json={'query': mutation, 'variables': variables},
      headers = {'Authorization': f"bearer {os.environ['TOKEN']}"}
    )

    json_response = json.loads(response.text)

    if 'errors' in json_response:
      logger.info(f"GraphQL error when adding {pr_id}: {json_response}")

def filter_prs(data, reviewer_id: str, project_id):
  """Given data about the draft state, reviewers, and project boards for PRs,
  return just the PRs that are:
  - not draft
  - are requesting a review for the specified team
  - are not already on the specified project board

  Arguments:
  data: A JSON object of this structure:
    {
      "data": {
      "node": {
        "pullRequests": {
          "nodes": [
              {
                "id": str, 
                "isDraft": bool, 
                "reviewRequests": {
                  "nodes": [
                    {
                      "requestedReviewer": {
                        "id": str
                      }
                    }...
                  ]
                }, 
                "projectCards": {
                  "nodes": [
                    {
                      "project": {
                        "id": str
                      }
                    }...
                  ]
                }
              }...
            ]
          }
        }
      }
    }
  reviewer_id: The node ID of the reviewer to filter for
  project_id: The project ID of the project to filter against

  Returns:
  A list of node IDs of the PRs that met the requirements
  """

  pr_data = data['data']['node']['pullRequests']['nodes']

  prs_to_add = []

  for pr in pr_data:
    if (
      not pr['isDraft'] and
      reviewer_id in [req_rev['requestedReviewer']['id'] for req_rev in pr['reviewRequests']['nodes'] if req_rev['requestedReviewer']] and
      project_id not in [proj_card['project']['id'] for proj_card in pr['projectCards']['nodes']]
    ):
      prs_to_add.append(pr['id'])
  
  return prs_to_add

def main():
  query_data = find_open_prs_for_repo(github_repo_id, num_prs_to_search)
  prs_to_add = filter_prs(query_data, docs_reviewers_id, docs_project_id)
  add_prs_to_board(prs_to_add, docs_column_id)

if __name__ == "__main__":
    main()
